### PR TITLE
[LoopRotate] Set loop back edge weight to not less than exit weight

### PR DESCRIPTION
**Pull request by HaohaiWen**
Branch weight from sample-based PGO may be not inaccurate due to
sampling. If the loop body must be executed, then origin loop back
edge weight must be not less than exit weight.

### CODE REVIEW

**Review by williamweixiao — APPROVED** [review]

**Comment by HaohaiWen** [comment]
Any more comments?

File: llvm/lib/Transforms/Utils/LoopRotationUtils.cpp [hunk 1]
```diff
@@ -347,9 +347,19 @@ static void updateBranchWeights(BranchInst &PreHeaderBI, BranchInst &LoopBI,
         // probabilities as if there are only 0-trip and 1-trip cases.
         ExitWeight0 = OrigLoopExitWeight - OrigLoopBackedgeWeight;
       }
+    } else {
+      if (OrigLoopExitWeight > OrigLoopBackedgeWeight) {
+        LLVM_DEBUG(
+            dbgs() << "WARNING: Bad loop back edge weight. Adjust it from "
+                   << OrigLoopBackedgeWeight << " to " << OrigLoopExitWeight
+                   << "\n");
```
Comment (WenleiHe) [inline llvm/lib/Transforms/Utils/LoopRotationUtils.cpp:355]: I don't think we need debug print warning for this. There is a lot of guessing and fixing already in the heuristic above, and it'd be inconsistent to just print debug warning for this case.

**Review by WenleiHe — COMMENTED** [review]
Thanks for the fix. I suggest you wait for @MatzeB to take a look.

**Review by WenleiHe — CHANGES_REQUESTED** [review]
Actually I'm a bit confused looking at the test case..

1) You can have zero-trip loop, in which case `loop_exit` can be reached without going into `loop_body`, why loop back edge needs to be not less than exit weight? 

2) In principle, profile inference for sample PGO should make input counts consistent. If there's input count inconsistency, we need track down the source of inconsistency. Fixing the input count inconsistency in a transformation isn't the right approach IMO. (The existing heuristic are making guesses for loop rotation which is different and is fine)

**Comment by MatzeB** [comment]
I don't follow why this would be a situation that needs fixup. While it is less common, it is perfectly fine for a loop to exit more often than using the loop backedge (when 0-trip counts happen often enough for a given loop to push the average trip count below 1).

**Comment by MatzeB** [comment]
Oh I see, this is for the case where the condition can be constant folded. I guess this makes sense in that case and you want to avoid an underflow? (may be good to mention in a comment/commit message).

**Comment by MatzeB** [comment]
I suspect some amount of noise in the sample / mapping samples to actual blocks etc. is unavoidable and I am not sure this situation (that only becomes "bad" when you consider the fact that the particular loop condition cannot be false in the first iteration) will be considered by smoothing algos. So seems fine to me, but leaving final call to @WenleiHe about whether some inaccuracies like this are considered normal (and hence must be worked around).

**Comment by HaohaiWen** [comment]
> Actually I'm a bit confused looking at the test case..
> 
> 1. You can have zero-trip loop, in which case `loop_exit` can be reached without going into `loop_body`, why loop back edge needs to be not less than exit weight?

Entering line 351 means HasConditionalPreHeader is false. That means this loop body must be executed at least once. Each execution of loop body must have a back edge to latch. Then it either go back to loop body or exit.


> 2. In principle, profile inference for sample PGO should make input counts consistent. If there's input count inconsistency, we need track down the source of inconsistency. Fixing the input count inconsistency in a transformation isn't the right approach IMO. (The existing heuristic are making guesses for loop rotation which is different and is fine)

Ideally, It should be fixed when loading profile in SampleProfileLoader. For this case, unless SampleProfilerLoader detecting the loop body must be executed at least (I think it's not easy for SampleProfileLoader), it have no idea whether this profile data is good.

**Comment by WenleiHe** [comment]
Ok, thanks for clarification. Please add a comment in the code explaining this case. Also clean up the debug prints.

File: llvm/lib/Transforms/Utils/LoopRotationUtils.cpp (hunk 1 above)
Comment (HaohaiWen) [reply llvm/lib/Transforms/Utils/LoopRotationUtils.cpp:355]: The heuristic wasn't based on guessing. I think those three expressions were bullet proofed (@MatzeB , please correct me if I am wrong). It just have a pre request the profile data is accurate. What I'm trying is to fix inaccurate profile data.
```
  //  -  x == x0 + x1        # counts to "exit" must stay the same.
  //  - y0 == x - x0 == x1   # how often loop was entered at all.
  //  - y1 == y - y0         # How often loop was repeated (after first iter.).
```

Comment (WenleiHe) [reply llvm/lib/Transforms/Utils/LoopRotationUtils.cpp:355]: > The heuristic wasn't based on guessing. I think those three expressions were bullet proofed

I was referring to the existing heuristic. With loop rotate, you cannot recover ground truth counts. And you have to make assumptions like "as if 0-trip count nearly never happens"/ "as if there are only 0-trip and 1-trip cases".

**Review by WenleiHe — APPROVED** [review]
The latest revision LGTM, thanks.

**Comment by alexfh** [comment]
We see regressions of up to 50% (depending on microarchitecture) in llvm-test-suite microbenchmarks, when compiled with PGO+thinlto:
* https://github.com/llvm/llvm-test-suite/blob/main/SingleSource/Benchmarks/Shootout/sieve.c ~ 13-50%
* https://github.com/llvm/llvm-test-suite/blob/main/SingleSource/Benchmarks/Linpack/linpack-pc.c ~ 13-22%
* https://github.com/llvm/llvm-test-suite/blob/main/SingleSource/Benchmarks/Stanford/Bubblesort.c ~ 15-20%

I don't think we consider these important, but I thought you might be interested in this observation.

**Pull request merged**